### ISSUE TITLE
Changing non-alphanumeric characters does not change the string

### ISSUE THREAD

**Issue by multiwebinc** [rainlab/translate-plugin#563]
Let's say, for example, I have the following string in one of my views:

```html
<p>{{ 'Choose one of the following:'|_ }}</p>
```

Then I later change my mind and decide to move the `:` outside the translation string:

```html
<p>{{ 'Choose one of the following'|_ }}:</p>
```

I then do a re-scan of the translate messages, however the string is still listed as `Choose one of the following:` with the `:` within the string, and when the translators see this, they will assume they need to include the `:` as well.

I see that in the `rainlab_translate_messages` table, the `code` column used looks like it removes all non-alphanumeric characters and replaces the spaces with a dot. Wouldn't it be more sensible to use a hash of the string instead of this? That way a single whitespace character change would trigger the whole string to be changed.

**Comment by mjauvin** [comment]
Yes, this is a known limitation. You need to delete the translation message and rescan.

What you suggest would wirk, but would cause major backward compatibility issues.

**Comment by multiwebinc** [comment]
@mjauvin What major backward-compatibility issues would it cause? Besides using it for an index to quickly find strings internally in the plugin, do other plugins somehow access these strings for another purpose?

**Comment by mjauvin** [comment]
@multiwebinc Anyone with the plugin currenlty installed would have major problems when updating the plugin to a version using a hash for the code since all their existing translations would be lost.

We'd need to do a migration but this could potentially create havoc.

**Comment by mjauvin** [comment]
FYI, I submitted a PR to allow overriding the method used to generate the messageCode.

ref, #564

**Comment by multiwebinc** [comment]
> We'd need to do a migration but this could potentially create havoc.

This is exactly what I was thinking as well. As long as no external plugins rely explicitly on the storage method used for the Translate plugin, run a migration on next update that simply updates all current codes to hashes.

Another option would be to select rows where the row is either a hash or the result of the current function. Of course, that adds slightly more computational power, but might be an option to allow the current code calculation method to be deprecated over several releases before it's finally removed.

**Comment by mjauvin** [comment]
@daftspunk what's your take on this, should we transition to md5 hash as we discussed before?

**Comment by daftspunk** [comment]
This has been fixed in October CMS v3.1 with a new translation key format.

- Link to the [upgrade guide](https://github.com/rainlab/translate-plugin/blob/master/UPGRADE.md)
- Link to [multisite documentation](https://docs.octobercms.com/3.x/cms/resources/multisite.html)

If you still need help, feel free to get in touch with the support desk for assistance: https://octobercms.com/contact

Thank you
